packages/firestore/src/local/lru_garbage_collector.ts

```diff
@@ -0,0 +1,268 @@
+  constructor(
+    private readonly capacity: number,
+    private readonly comparator: (a: T, b: T) => number
+  ) {
+    assert(capacity > 0, 'Capacity must be greater than 0');
+    this.queue = new Array<T>(capacity);
+  }
+
+  add(elem: T): void {
+    assert(this._size + 1 <= this.capacity, 'Queue is over capacity');
+    if (this._size === 0) {
+      this.queue[0] = elem;
+      this._size = 1;
+    } else {
+      this.siftUp(elem);
+    }
+  }
+
+  poll(): T | null {
+    if (this._size === 0) {
+      return null;
+    }
+    const result = this.queue[0];
+    const newSize = --this._size;
+    const last = this.queue[newSize];
+    delete this.queue[newSize];
+    if (newSize !== 0) {
+      this.siftDown(last);
+    }
+    return result;
+  }
+
+  peek(): T | null {
+    if (this._size > 0) {
+      return this.queue[0];
+    }
+    return null;
+  }
+
+  private siftUp(elem: T): void {
+    let k = this._size;
+    while (k > 0) {
+      const parent = (k - 1) >>> 1;
+      const toCheck = this.queue[parent];
+      const comp = this.comparator(elem, toCheck);
+      if (comp >= 0) {
+        break;
+      }
+      this.queue[k] = toCheck;
+      k = parent;
+    }
+    this.queue[k] = elem;
+    this._size++;
+  }
+
+  private siftDown(lastElem: T): void {
+    let k = 0;
+    const half = this._size >>> 1;
+    while (k < half) {
+      let child = (k << 1) + 1;
+      let toCheck = this.queue[child];
+      const right = child + 1;
+      if (
+        right < this._size &&
+        this.comparator(toCheck, this.queue[right]) > 0
+      ) {
+        toCheck = this.queue[right];
+        child = right;
+      }
+      if (this.comparator(lastElem, toCheck) <= 0) {
+        break;
+      }
+      this.queue[k] = toCheck;
+      k = child;
+    }
+    this.queue[k] = lastElem;
+  }
+}
+
+/**
+ * Used to calculate the nth sequence number. Keeps a rolling buffer of the lowest n values passed
+ * to `addElement`, and finally reports the largest of them in `maxValue`.
+ */
+class RollingSequenceNumberBuffer {
+  private queue: PriorityQueue<ListenSequenceNumber>;
+
+  // Invert the comparison because we want to keep the smallest values.
+  private static COMPARATOR: (
+    a: ListenSequenceNumber,
+    b: ListenSequenceNumber
+  ) => number = (a, b) => {
+    if (b < a) {
+      return -1;
+    } else if (b === a) {
+      return 0;
+    }
+    return 1;
+  };
+
+  constructor(private readonly maxElements: number) {
+    this.queue = new PriorityQueue(
+      maxElements,
+      RollingSequenceNumberBuffer.COMPARATOR
+    );
+  }
+
+  addElement(sequenceNumber: ListenSequenceNumber): void {
+    if (this.queue.size < this.maxElements) {
+      this.queue.add(sequenceNumber);
+    } else {
+      // Note: use first because we have inverted the comparison
+      const highestValue = this.queue.peek()!;
+      if (sequenceNumber < highestValue) {
+        this.queue.poll();
+        this.queue.add(sequenceNumber);
+      }
+    }
```

Comment:
As currently written, that won't work as we may go over capacity on the priority queue. I did not want to port the code to grow the underlying array since we already know the max size we want when we construct it.

We could make this change if we switch out the priority queue to something else.